sympy/core/function.py

```diff
@@ -1105,8 +1105,9 @@ def __new__(cls, expr, *variables, **assumptions):
         # functions and Derivatives as those can be created by intermediate
         # derivatives.
         if evaluate:
-            symbol_set = set(sc[0] for sc in variable_count if sc[0].is_Symbol)
-            if symbol_set.difference(expr.free_symbols):
+            from sympy import IndexedBase
+            symbol_set = set(sc[0].base if sc[0].is_Indexed else sc[0] for sc in variable_count if sc[0].is_Symbol)
+            if symbol_set.difference(expr.free_symbols).difference(expr.atoms(IndexedBase)):
```

Comment:
@asmeurer this was the trick to allow `A[i].diff(A[j])`. This is an optimization, it checks whether the deriving varible is contained in the expression, otherwise return zero (i.e. deriving a constant).